auth/impersonate/user.go

```diff
@@ -0,0 +1,180 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package impersonate
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+
+	"cloud.google.com/go/auth"
+	"cloud.google.com/go/auth/internal"
+)
+
+func user(opts *CredentialOptions, client *http.Client, lifetime time.Duration, isStaticToken bool) (auth.TokenProvider, error) {
+	u := userTokenProvider{
+		client:          client,
+		targetPrincipal: opts.TargetPrincipal,
+		subject:         opts.Subject,
+		lifetime:        lifetime,
+	}
+	u.delegates = make([]string, len(opts.Delegates))
+	for i, v := range opts.Delegates {
+		u.delegates[i] = formatIAMServiceAccountName(v)
+	}
+	u.scopes = make([]string, len(opts.Scopes))
+	copy(u.scopes, opts.Scopes)
+	if isStaticToken {
+		return auth.NewCachedTokenProvider(u, &auth.CachedTokenProviderOptions{
+			DisableAutoRefresh: true,
+		}), nil
+	}
+	return auth.NewCachedTokenProvider(u, nil), nil
+}
+
+type claimSet struct {
+	Iss   string `json:"iss"`
+	Scope string `json:"scope,omitempty"`
+	Sub   string `json:"sub,omitempty"`
+	Aud   string `json:"aud"`
+	Iat   int64  `json:"iat"`
+	Exp   int64  `json:"exp"`
+}
+
+type signJWTRequest struct {
+	Payload   string   `json:"payload"`
+	Delegates []string `json:"delegates,omitempty"`
+}
+
+type signJWTResponse struct {
+	// KeyID is the key used to sign the JWT.
+	KeyID string `json:"keyId"`
+	// SignedJwt contains the automatically generated header; the
+	// client-supplied payload; and the signature, which is generated using
+	// the key referenced by the `kid` field in the header.
+	SignedJWT string `json:"signedJwt"`
+}
+
+type exchangeTokenResponse struct {
+	AccessToken string `json:"access_token"`
+	TokenType   string `json:"token_type"`
+	ExpiresIn   int64  `json:"expires_in"`
+}
+
+type userTokenProvider struct {
+	client *http.Client
+
+	targetPrincipal string
+	subject         string
+	scopes          []string
+	lifetime        time.Duration
+	delegates       []string
+}
+
+func (u userTokenProvider) Token(ctx context.Context) (*auth.Token, error) {
+	signedJWT, err := u.signJWT()
+	if err != nil {
+		return nil, err
+	}
+	return u.exchangeToken(ctx, signedJWT)
+}
+
+func (u userTokenProvider) signJWT() (string, error) {
+	now := time.Now()
+	exp := now.Add(u.lifetime)
+	claims := claimSet{
+		Iss:   u.targetPrincipal,
+		Scope: strings.Join(u.scopes, " "),
+		Sub:   u.subject,
+		Aud:   fmt.Sprintf("%s/token", oauth2Endpoint),
+		Iat:   now.Unix(),
+		Exp:   exp.Unix(),
+	}
+	payloadBytes, err := json.Marshal(claims)
+	if err != nil {
+		return "", fmt.Errorf("impersonate: unable to marshal claims: %v", err)
```

Comment:
```suggestion
		return "", fmt.Errorf("impersonate: unable to marshal claims: %w", err)
```
Maybe wrap, here and elsewhere.